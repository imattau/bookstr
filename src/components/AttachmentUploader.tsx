import React, { useState } from 'react';
import { useNostr } from '../nostr';

async function uploadAttachment(file: File): Promise<string> {
  const fd = new FormData();
  fd.append('file', file);
  const res = await fetch('/api/upload', {
    method: 'POST',
    body: fd,
  });
  if (!res.ok) throw new Error('upload failed');
  const data = await res.json();
  return data.url as string;
}

interface AttachmentUploaderProps {
  bookId?: string;
  onUploaded?: (url: string) => void;
}

export const AttachmentUploader: React.FC<AttachmentUploaderProps> = ({
  bookId,
  onUploaded,
}) => {
  const ctx = useNostr();
  const [loading, setLoading] = useState(false);

  const handleFile = async (
    e: React.ChangeEvent<HTMLInputElement>,
  ) => {
    const file = e.target.files?.[0];
    if (!file) return;
    setLoading(true);
    try {
      const url = await uploadAttachment(file);
      await ctx.publish({
        kind: 1064,
        content: '',
        tags: [
          ['mime', file.type],
          ['url', url],
          ...(bookId ? [['e', bookId]] : []),
        ],
      });
      onUploaded?.(url);
    } catch {
      // ignore errors
    } finally {
      setLoading(false);
      e.target.value = '';
    }
  };

  return (
    <div className="space-y-2">
      <input type="file" onChange={handleFile} disabled={loading} />
      {loading && <p className="text-sm">Uploading...</p>}
    </div>
  );
};
