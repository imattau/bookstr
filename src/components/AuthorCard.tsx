import React from 'react';
import { FollowButton } from './FollowButton';

export interface AuthorCardProps {
  pubkey: string;
  name?: string;
}

/**
 * Simple author card showing a name/pubkey with follow button.
 */
export const AuthorCard: React.FC<AuthorCardProps> = ({ pubkey, name }) => {
  return (
    <div className="flex items-center gap-[var(--space-2)] rounded-[var(--radius-card)] border p-[var(--space-2)]">
      <div className="flex-1">
        <p className="font-medium">{name || pubkey}</p>
      </div>
      <FollowButton pubkey={pubkey} />
    </div>
  );
};
