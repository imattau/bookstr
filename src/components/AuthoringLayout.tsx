import React from 'react';

export interface AuthoringLayoutProps {
  sidebar: React.ReactNode;
  content: React.ReactNode;
}

/**
 * Layout used for authoring screens with optional sidebar.
 */
export const AuthoringLayout: React.FC<AuthoringLayoutProps> = ({ sidebar, content }) => {
  return (
    <div className="mx-auto max-w-5xl px-4 py-6">
      <div className="grid gap-6 lg:grid-cols-[minmax(12rem,1fr)_3fr]">
        <aside className="hidden lg:block">{sidebar}</aside>
        <main>{content}</main>
      </div>
    </div>
  );
};
