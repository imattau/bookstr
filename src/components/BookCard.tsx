import React, { useState } from 'react';
import { useNostr, zap } from '../nostr';
import type { Event as NostrEvent } from 'nostr-tools';
import { logEvent } from '../analytics';

interface BookCardProps {
  event: NostrEvent;
}

export const BookCard: React.FC<BookCardProps> = ({ event }) => {
  const title = event.tags.find((t) => t[0] === 'title')?.[1] ?? 'Untitled';
  const summary = event.tags.find((t) => t[0] === 'summary')?.[1] ?? '';
  const cover = event.tags.find((t) => t[0] === 'image')?.[1];
  const ctx = useNostr();
  const [status, setStatus] = useState<'idle' | 'zapping' | 'done'>('idle');
  const { toggleBookmark, bookmarks } = ctx;

  const handleZap = async () => {
    setStatus('zapping');
    try {
      await zap(ctx, event);
      logEvent('click_vote', { id: event.id });
      setStatus('done');
    } catch {
      setStatus('idle');
    }
  };

  const handleFav = async () => {
    try {
      await toggleBookmark(event.id);
      logEvent('click_fav', { id: event.id });
    } catch {
      /* ignore */
    }
  };

  return (
    <div className="rounded border p-2">
      {cover && (
        <img src={cover} alt="" className="mb-2 h-32 w-24 object-cover" />
      )}
      <h3 className="font-semibold">{title}</h3>
      {summary && <p className="text-sm text-gray-500">{summary}</p>}
      <div className="pt-2 flex gap-2">
        <button
          onClick={handleZap}
          className="rounded bg-yellow-500 px-2 py-1 text-white"
        >
          {status === 'zapping'
            ? 'Zapping...'
            : status === 'done'
              ? 'Zapped!'
              : 'Zap'}
        </button>
        <button
          onClick={handleFav}
          aria-label="Favorite"
          className="rounded border px-2 py-1"
        >
          â˜…
        </button>
      </div>
    </div>
  );
};
