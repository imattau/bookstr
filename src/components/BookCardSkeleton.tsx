import React from 'react';
import { Skeleton, TextSkeleton } from './Skeleton';

export const BookCardSkeleton: React.FC = () => (
  <div className="rounded-[var(--radius-card)] border p-[var(--space-2)] space-y-[var(--space-2)]">
    <Skeleton className="h-32 w-24" />
    <TextSkeleton lines={1} className="w-3/4" />
    <TextSkeleton lines={1} />
    <div className="pt-2 flex gap-2">
      <Skeleton className="h-6 w-16 rounded" />
      <Skeleton className="h-6 w-8 rounded" />
    </div>
  </div>
);
