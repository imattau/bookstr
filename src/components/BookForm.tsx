import React from 'react';
import { useNostr } from '../nostr';

export const BookForm: React.FC = () => {
  const { publish } = useNostr();
  const [title, setTitle] = React.useState('');
  const [content, setContent] = React.useState('');
  const [summary, setSummary] = React.useState('');
  const [tags, setTags] = React.useState('');
  const [cover, setCover] = React.useState('');

  const handlePublish = async () => {
    await publish({
      kind: 30023,
      content,
      tags: [
        ['title', title],
        ['summary', summary],
        ...tags
          .split(',')
          .map((t) => t.trim())
          .filter(Boolean)
          .map((t) => ['t', t]),
        ...(cover ? [['image', cover]] : []),
      ],
    });
    setTitle('');
    setSummary('');
    setContent('');
    setTags('');
    setCover('');
  };

  return (
    <div className="space-y-2">
      <input
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        placeholder="Title"
        className="w-full rounded border p-2"
      />
      <textarea
        value={summary}
        onChange={(e) => setSummary(e.target.value)}
        placeholder="Summary"
        className="w-full rounded border p-2"
      />
      <textarea
        value={content}
        onChange={(e) => setContent(e.target.value)}
        placeholder="Content"
        className="w-full rounded border p-2"
      />
      <input
        value={tags}
        onChange={(e) => setTags(e.target.value)}
        placeholder="Tags comma separated"
        className="w-full rounded border p-2"
      />
      <input
        value={cover}
        onChange={(e) => setCover(e.target.value)}
        placeholder="Cover URL"
        className="w-full rounded border p-2"
      />
      <button
        onClick={handlePublish}
        className="rounded bg-primary-600 px-4 py-2 text-white"
      >
        Publish
      </button>
    </div>
  );
};
