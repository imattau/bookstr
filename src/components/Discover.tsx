import React, { useEffect, useRef, useState } from 'react';
import type { Event as NostrEvent, Filter } from 'nostr-tools';
import { useNostr } from '../nostr';
import { BookCard } from './BookCard';
import { useOnboarding } from '../useOnboarding';
import { logEvent } from '../analytics';

const TAGS = ['All', 'Fiction', 'Mystery', 'Fantasy'];

export const Discover: React.FC = () => {
  const { subscribe, contacts } = useNostr();
  const [events, setEvents] = useState<NostrEvent[]>([]);
  const [votes, setVotes] = useState<Record<string, number>>({});
  const voteIds = useRef(new Set<string>());
  const [search, setSearch] = useState('');
  const [tag, setTag] = useState('All');

  useEffect(() => {
    logEvent('discover_view');
  }, []);

  useEffect(() => {
    const filters: Filter[] = [{ kinds: [30023], limit: 50 }];
    if (contacts.length) filters[0].authors = contacts;
    const off = subscribe(filters, (evt) =>
      setEvents((e) => {
        if (e.find((x) => x.id === evt.id)) return e;
        return [...e, evt];
      }),
    );
    return off;
  }, [subscribe, contacts]);

  // aggregate votes for known events
  useEffect(() => {
    if (!events.length) return;
    const ids = events.map((e) => e.id);
    const off = subscribe([{ kinds: [7], '#e': ids }], (evt) => {
      if (voteIds.current.has(evt.id)) return;
      voteIds.current.add(evt.id);
      const target = evt.tags.find((t) => t[0] === 'e')?.[1];
      if (target) {
        setVotes((v) => ({ ...v, [target]: (v[target] ?? 0) + 1 }));
      }
    });
    return off;
  }, [events, subscribe]);

  const filtered = events.filter((evt) => {
    let ok = true;
    if (contacts.length && !contacts.includes(evt.pubkey)) ok = false;
    if (tag !== 'All') {
      ok = evt.tags.some(
        (t) => t[0] === 't' && t[1].toLowerCase() === tag.toLowerCase(),
      );
    }
    if (ok && search) {
      const title = evt.tags.find((t) => t[0] === 'title')?.[1] ?? '';
      ok = title.toLowerCase().includes(search.toLowerCase());
    }
    return ok;
  });

  const topRated = [...events]
    .sort((a, b) => (votes[b.id] ?? 0) - (votes[a.id] ?? 0))
    .slice(0, 6);

  const newReleases = [...events]
    .sort((a, b) => (b.created_at ?? 0) - (a.created_at ?? 0))
    .slice(0, 6);

  const recommended = filtered.slice(0, 6);

  const searchOnboarding = useOnboarding('discover-search', 'Search for books');

  return (
    <div className="pb-4">
      <header className="relative bg-primary-600 py-3 text-center text-white">
        <button
          ref={searchOnboarding.ref as React.RefObject<HTMLButtonElement>}
          onClick={searchOnboarding.dismiss}
          aria-label="Search"
          className={`relative absolute left-2 top-1/2 -translate-y-1/2 ${
            searchOnboarding.show ? 'rounded ring-2 ring-primary-300' : ''
          }`}
        >
          üîç
          {searchOnboarding.Tooltip}
        </button>
        <h1 className="text-[18px] font-semibold">Bookstr</h1>
      </header>
      <div className="p-4">
        <input
          value={search}
          onChange={(e) => setSearch(e.target.value)}
          placeholder="Search"
          className="w-full rounded border border-border p-2 focus-visible:outline-none focus-visible:[box-shadow:var(--focus-ring)]"
        />
      </div>
      <div className="flex overflow-x-auto gap-1 px-2 pb-2">
        {TAGS.map((t) => (
          <button
            key={t}
            onClick={() => setTag(t)}
            className={`btn-tap whitespace-nowrap rounded px-2 py-1 text-[14px] ${
              tag === t
                ? 'bg-primary-600 text-white'
                : 'border border-border text-primary-600 bg-[color:var(--clr-surface)]'
            }`}
          >
            {t}
          </button>
        ))}
      </div>
      <section className="p-4">
        <h2 className="mb-2 font-semibold">Top Rated</h2>
        <div className="grid grid-cols-2 gap-4">
          {topRated.map((e) => (
            <BookCard key={e.id} event={e as NostrEvent} />
          ))}
        </div>
      </section>
      <section className="p-4">
        <h2 className="mb-2 font-semibold">New Releases</h2>
        <div className="grid grid-cols-2 gap-4">
          {newReleases.map((e) => (
            <BookCard key={e.id} event={e as NostrEvent} />
          ))}
        </div>
      </section>
      <section className="p-4">
        <h2 className="mb-2 font-semibold">Recommended for You</h2>
        <div className="grid grid-cols-2 gap-4">
          {recommended.map((e) => (
            <BookCard key={e.id} event={e as NostrEvent} />
          ))}
        </div>
      </section>
    </div>
  );
};
