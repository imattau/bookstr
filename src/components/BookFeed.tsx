import React, { useEffect, useState } from 'react';
import { useNostr } from '../nostr';
import { BookCard } from './BookCard';
import { BookCardSkeleton } from './BookCardSkeleton';
import type { Event as NostrEvent } from 'nostr-tools';
import { useEventStore } from '../store/events';

/**
 * Feed showing recent book publications from followed relays.
 */
export const BookFeed: React.FC = () => {
  const { subscribe } = useNostr();
  const addEvent = useEventStore((s) => s.addEvent);
  const [events, setEvents] = useState<
    (NostrEvent & { repostedBy?: string })[]
  >([]);

  useEffect(() => {
    const offMain = subscribe([{ kinds: [30023], limit: 20 }], (evt) => {
      addEvent(evt);
      setEvents((e) => {
        const d = evt.tags.find((t) => t[0] === 'd')?.[1];
        if (d) {
          const idx = e.findIndex(
            (x) =>
              x.kind === 30023 &&
              x.pubkey === evt.pubkey &&
              x.tags.find((t) => t[0] === 'd')?.[1] === d,
          );
          if (idx !== -1) {
            if (e[idx].created_at >= evt.created_at) return e;
            const copy = [...e];
            copy[idx] = evt;
            return copy;
          }
        }
        if (e.find((x) => x.id === evt.id)) return e;
        return [...e, evt];
      });
    });
    const offRepost = subscribe([{ kinds: [6], limit: 20 }], (evt) => {
      const target = evt.tags.find((t) => t[0] === 'e')?.[1];
      if (!target) return;
      const offTarget = subscribe([{ ids: [target] }], (orig) => {
        addEvent(orig);
        setEvents((e) => {
          const d = orig.tags.find((t) => t[0] === 'd')?.[1];
          if (d) {
            const idx = e.findIndex(
              (x) =>
                x.kind === 30023 &&
                x.pubkey === orig.pubkey &&
                x.tags.find((t) => t[0] === 'd')?.[1] === d,
            );
            if (idx !== -1) {
              if (e[idx].created_at >= orig.created_at) {
                const copy = [...e];
                copy[idx] = { ...copy[idx], repostedBy: evt.pubkey };
                return copy;
              }
              const copy = [...e];
              copy[idx] = { ...orig, repostedBy: evt.pubkey };
              return copy;
            }
          }
          const idx = e.findIndex((x) => x.id === orig.id);
          if (idx !== -1) {
            const copy = [...e];
            copy[idx] = { ...copy[idx], repostedBy: evt.pubkey };
            return copy;
          }
          return [...e, { ...orig, repostedBy: evt.pubkey }];
        });
        offTarget();
      });
    });
    return () => {
      offMain();
      offRepost();
    };
  }, [subscribe, addEvent]);

  return (
    <ul role="list" className="space-y-4">
      {events.length === 0
        ? Array.from({ length: 3 }).map((_, i) => (
            <li key={i} role="listitem">
              <BookCardSkeleton />
            </li>
          ))
        : events.map((e) => (
            <li key={e.id} role="listitem">
              <BookCard
                event={e}
                onDelete={(id) =>
                  setEvents((evts) => evts.filter((x) => x.id !== id))
                }
              />
            </li>
          ))}
    </ul>
  );
};
