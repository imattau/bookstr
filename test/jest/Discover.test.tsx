import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Discover } from '../../src/components/Discover';

jest.mock('../../src/analytics', () => ({ logEvent: jest.fn() }));

jest.mock('../../src/nostr', () => ({
  useNostr: () => ({ contacts: [], subscribe: jest.fn() }),
}));

jest.mock('../../src/hooks/useDiscoverBooks', () => ({
  useDiscoverBooks: () => ({
    books: [],
    trending: [],
    newReleases: [],
    loading: false,
    removeBook: jest.fn(),
  }),
}));

describe('Discover', () => {
  it('mobile snapshot', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 375 });
    window.dispatchEvent(new Event('resize'));
    const { container } = render(
      <MemoryRouter>
        <Discover />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('desktop snapshot', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1200 });
    window.dispatchEvent(new Event('resize'));
    const { container } = render(
      <MemoryRouter>
        <Discover />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
