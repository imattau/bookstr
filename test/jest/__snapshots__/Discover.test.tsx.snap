// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Discover desktop snapshot 1`] = `
<div
  class="pb-[var(--space-4)]"
>
  <header
    class="flex items-center gap-[var(--space-2)] bg-[color:var(--clr-primary-600)] p-[var(--space-3)]"
  >
    <h1
      class="text-[18px] font-semibold text-white"
    >
      Bookstr
    </h1>
    <div
      class="flex flex-1 items-center gap-1 relative rounded ring-2 ring-primary-300"
    >
      <input
        class="w-full rounded-[var(--radius-button)] border border-border p-[var(--space-2)] bg-[color:var(--clr-surface)] text-[color:var(--clr-text)] focus-visible:outline-none focus-visible:[box-shadow:var(--focus-ring)]"
        placeholder="Search"
        value=""
      />
      <div
        class="pointer-events-none absolute z-10 mt-[var(--space-1)] whitespace-nowrap rounded bg-primary-600 px-2 py-1 text-[12px] text-white shadow"
      >
        Search for books
      </div>
    </div>
  </header>
  <div
    class="flex overflow-x-auto gap-[var(--space-2)] px-[var(--space-2)] pb-[var(--space-2)]"
  >
    <button
      aria-label="All"
      aria-pressed="true"
      class="btn-tap whitespace-nowrap rounded-[var(--radius-button)] px-[var(--space-2)] py-[var(--space-1)] text-[14px] bg-[color:var(--clr-primary-600)] text-white"
    >
      All
    </button>
  </div>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Trending Books
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      New Releases
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Recommended for You
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Communities
    </h2>
    <ul
      class="space-y-4"
      role="list"
    />
  </section>
</div>
`;

exports[`Discover mobile snapshot 1`] = `
<div
  class="pb-[var(--space-4)]"
>
  <header
    class="flex items-center gap-[var(--space-2)] bg-[color:var(--clr-primary-600)] p-[var(--space-3)]"
  >
    <h1
      class="text-[18px] font-semibold text-white"
    >
      Bookstr
    </h1>
    <div
      class="flex flex-1 items-center gap-1 relative rounded ring-2 ring-primary-300"
    >
      <input
        class="w-full rounded-[var(--radius-button)] border border-border p-[var(--space-2)] bg-[color:var(--clr-surface)] text-[color:var(--clr-text)] focus-visible:outline-none focus-visible:[box-shadow:var(--focus-ring)]"
        placeholder="Search"
        value=""
      />
      <div
        class="pointer-events-none absolute z-10 mt-[var(--space-1)] whitespace-nowrap rounded bg-primary-600 px-2 py-1 text-[12px] text-white shadow"
      >
        Search for books
      </div>
    </div>
  </header>
  <div
    class="flex overflow-x-auto gap-[var(--space-2)] px-[var(--space-2)] pb-[var(--space-2)]"
  >
    <button
      aria-label="All"
      aria-pressed="true"
      class="btn-tap whitespace-nowrap rounded-[var(--radius-button)] px-[var(--space-2)] py-[var(--space-1)] text-[14px] bg-[color:var(--clr-primary-600)] text-white"
    >
      All
    </button>
  </div>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Trending Books
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      New Releases
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Recommended for You
    </h2>
    <ul
      class="grid grid-cols-1 gap-[var(--space-4)] md:grid-cols-2 lg:grid-cols-4"
      role="list"
    />
  </section>
  <section
    class="p-[var(--space-4)]"
  >
    <h2
      class="mb-[var(--space-2)] font-semibold"
    >
      Communities
    </h2>
    <ul
      class="space-y-4"
      role="list"
    />
  </section>
</div>
`;
